### ISSUE TITLE
Show frames instead of milliseconds

### ISSUE THREAD

**Issue by GHubbler** [iina/iina#3153]
For Audio it might be ok to have precise time display with milliseconds.
For movie files it would be more practicable to have the current frame displayed after the seconds, like in video editing programs usable: h:mm:ss:CurrentFrame.

**Comment by theonlygusti** [comment]
How did you get it show milliseconds?

**Comment by GHubbler** [comment]
Right mouse click one of the time-counters in the controls. 😉 
![MS_IINA](https://user-images.githubusercontent.com/10322831/104379054-32d87a00-5529-11eb-9dca-a90a21278e1c.png)

**Comment by wantg** [comment]
WF........
How did you know that

**Comment by patg75** [comment]
> For Audio it might be ok to have precise time display with milliseconds. For movie files it would be more practicable to have the current frame displayed after the seconds, like in video editing programs usable: h:mm:ss:CurrentFrame.

This! I'd love to have **both** frame count **AND** timecode h:mm:ss:ff as display options. So inconvenient to have to open clips in a second app to find out the TC or current frame.

To add to this: it would be perfect if saving a still frame could optionally include TC or frame as an appendix in the file name instead of the counter number.

**Comment by ddofborg** [comment]
+1 for frame an option. Or can we do this already?

**Comment by Globolo001** [comment]
Agreed. Please do this.

Preferably also with a seconds:framenumber

Milliseconds are practically meaningless with inherently quantized timeframes.

**Comment by pindab0ter** [comment]
I would also find this very useful.

**Comment by yorkyang2333** [comment]
I've opened PR #6256 in order to implement this, please check it out!

**Comment by patg75** [comment]
Sorry, this is a long one:

I tested this with a NTSC file (24fps Big Buck Bunny re-encoded to `30000/1001` via ffmpeg) to check the concern about fractional frame rates over long durations.

If I understand this correctly, the frame calculation doesn't actually accumulate drift over time, because it's stateless:

```swift
frame = floor((second - floor(second)) * frameRate)
```

This recomputes the frame index from the absolute playback time on every call. There's no running frame counter being incremented across calls. So the classic NTSC drift problem (where a 30fps-assumption clock diverges from wall-clock time over hours) doesn't apply here. Each display update is an independent snapshot, not an accumulation. I verified this at various positions (e.g. seeking to `9:30.5` on a 29.97fps file correctly shows frame `15` = `round(0.5 * 30000/1001)`).

A few things worth pointing out though:

1. Big Buck Bunny is too short for real drift tests but the code should hold up because of its stateless design.

2. As mentioned before, this isn't SMPTE timecode, and I don't think it needs to be! Sure, it doesn't have drop-frame handling, but as long as this is stated as a "current frame within this second" indicator and not a broadcast TC, all is well IMO. To set that expectation it would be worth a comment in the help docs on `stringRepresentationWithPrecision`, since "1 frame" precision could otherwise read as timecode-accurate to some users. Even better if there's a direct GUI label that could be used for said comment.

3. The real accuracy depends on how closely `second` (I assume this is mpv's `time-pos`) tracks the actual decoded frame's PTS. Since `time-pos` is an interpolated/reported position rather than guaranteed to be exactly the current frame's PTS, there's a theoretical off-by-one risk right at frame boundaries (e.g. right after pausing, or with VFR sources where `demuxFps` doesn't reflect the actual per-frame timing). I haven't been able to reproduce a visible off-by-one in testing, but it might be worth a quick check with frame-stepping near a boundary to confirm the field always agrees with the actual current frame, not the previous one.

4. **Suggestion:** mpv already gives us `estimated-frame-number`, which tracks the actual decoder-side frame index rather than deriving it from `time-pos` + nominal `demuxFps`.

   4.1. Interestingly, IINA already declares this property (`MPVProperty.estimatedFrameNumber` in `MPVProperty.swift`) but it isn't used anywhere in the codebase yet it seems?! Reading the frame index from mpv directly instead of recomputing it from time would sidestep both the VFR case (where `demuxFps` is meaningless) and any transient mismatch between `time-pos` and the actually displayed frame right after a seek/pause.

   4.2. That would also open the door to a feature I'd find genuinely useful and mentioned in my request: an option to display the absolute frame count (a single running number since the start of the file) rather than only the `HH:MM:SS:FF`-style (pseudo)timecode this PR adds. For frame accurate editing/reference work, jumping to or reading off an absolute frame number is often more directly useful than a per second frame index, and `estimated-frame-number` gives you that for free without any of the fractional second math.

   4.3. If both options end up in the same menu, maybe even adding a visual separator line between the time precision entries (`1 second` – `1 millisecond`) and the frame based ones (`1 frame` + `frame count`), since they're conceptually different. One's a display resolution, the other's a display MODE. Not a big deal either way, just flagging it in case it helps.

tl;dr I don't think fractional framerate drift is a blocker for this since the implementation sidesteps that specific issue by design.

Very happy to help test, if useful.

**Comment by low-batt** [comment]
On this side issue:
> Interestingly, IINA already declares this property (MPVProperty.estimatedFrameNumber in MPVProperty.swift) but it isn't used anywhere in the codebase yet it seems?!

The answer to why IINA is declaring unused properties is found in the documentation comment for the `MPVProperty` class:
https://github.com/iina/iina/blob/d1dca0a5101f31ad31e8aea854a15f963ea58f97/iina/MPVProperty.swift#L3-L11

In Xcode this class is found in a group called `Generated` along with `MPVCommand` and 	`MPVOption`.

The `other/parse_doc.rb` script that generates these source files is a bit iffy as it is parsing the mpv manual to gather the information needed for generation. Frequently mpv changes the manual in some way that confuses the generator and the script has to be updated. This is one of the reasons the generated sources are checked into git instead of being part of the build.
